bpfman/src/lib.rs

```diff
@@ -460,12 +460,14 @@ pub fn attach_program(
     let name = prog.get_data().get_name().unwrap_or("not set".to_string());
     info!("Request to attach {kind} program named \"{name}\" with id {id}");
 
-    // Write attach info into the database
+    // Write attach info into the database. Once written to the database,
+    // DO NOT EXIT with a failure without calling prog.remove_link().
     let mut link = prog.add_link()?;
-    link.attach(attach_info)?;
-
-    let result = attach_program_internal(config, root_db, &prog, link.clone());
 
+    let result = match link.attach(attach_info) {
+        Ok(_) => attach_program_internal(config, root_db, &prog, link.clone()),
+        Err(e) => Err(e),
+    };
```

Comment:
```suggestion
let result = link.attach(attach_info)
    .and_then(|_| attach_program_internal(config, root_db, &prog, link.clone()));
```
Just a nit.  What you have works.  I think this would work too and would be a little cleaner.  Your choice.